runtime/apps/pldm/pldm-lib/src/control_context.rs

```diff
@@ -0,0 +1,803 @@
+        supported_commands: &'a [CmdOpCode],
+    ) -> Result<Self, PldmError> {
+        Ok(Self {
+            pldm_type,
+            protocol_version: match PldmVersion::try_from(protocol_version) {
+                Ok(ver) => ver.bcd_encode_to_ver32(),
+                Err(_) => return Err(PldmError::InvalidProtocolVersion),
+            },
+            supported_commands,
+        })
+    }
+}
+
+/// `ControlContext` is a structure that holds the control context for the PLDM library.
+///
+/// # Fields
+///
+/// * `tid` - An atomic unsigned size integer representing the transaction ID.
+/// * `capabilities` - A reference to a slice of `ProtocolCapability` which represents the protocol capabilities.
+pub struct ControlContext<'a> {
+    tid: AtomicUsize,
+    capabilities: &'a [ProtocolCapability<'a>],
+}
+
+impl<'a> ControlContext<'a> {
+    pub fn new(capabilities: &'a [ProtocolCapability<'a>]) -> Self {
+        Self {
+            tid: AtomicUsize::new(UNASSIGNED_TID as usize),
+            capabilities,
+        }
+    }
+
+    pub fn get_tid(&self) -> Tid {
+        self.tid.load(Ordering::SeqCst) as Tid
+    }
+
+    pub fn set_tid(&self, tid: Tid) {
+        self.tid.store(tid as usize, Ordering::SeqCst);
+    }
+
+    pub fn get_supported_commands(
+        &self,
+        pldm_type: PldmSupportedType,
+        protocol_version: Ver32,
+    ) -> Option<&[CmdOpCode]> {
+        self.capabilities
+            .iter()
+            .find(|cap| cap.pldm_type == pldm_type && cap.protocol_version == protocol_version)
+            .map(|cap| cap.supported_commands)
+    }
+
+    pub fn get_protocol_versions(
+        &self,
+        pldm_type: PldmSupportedType,
+        versions: &mut [Ver32],
+    ) -> usize {
+        let mut count = 0;
+        for cap in self
+            .capabilities
+            .iter()
+            .filter(|cap| cap.pldm_type == pldm_type)
+        {
+            if count < versions.len() {
+                versions[count] = cap.protocol_version;
+                count += 1;
+            } else {
+                break;
+            }
+        }
+        count
+    }
+
+    pub fn get_supported_types(&self, types: &mut [u8]) -> usize {
+        let mut count = 0;
+        for cap in self.capabilities.iter() {
+            let pldm_type = cap.pldm_type as u8;
+            if !types[..count].contains(&pldm_type) {
+                if count < types.len() {
+                    types[count] = pldm_type;
+                    count += 1;
+                } else {
+                    break;
+                }
+            }
+        }
+        count
+    }
+
+    pub fn is_supported_type(&self, pldm_type: PldmSupportedType) -> bool {
+        self.capabilities
+            .iter()
+            .any(|cap| cap.pldm_type == pldm_type)
+    }
+
+    pub fn is_supported_version(
+        &self,
+        pldm_type: PldmSupportedType,
+        protocol_version: Ver32,
+    ) -> bool {
+        let mut versions = [0; 4];
```

Comment:
is there a maximum number of versions or capabilities we could use as a `const` for these arrays, and use that `const` everywhere?